composer/utils/object_store/oci_object_store.py

```diff
@@ -0,0 +1,104 @@
+
+"""S3-Compatible object store."""
+
+from __future__ import annotations
+
+import os
+import pathlib
+import uuid
+from typing import Callable, Optional, Union
+
+from composer.utils.import_helpers import MissingConditionalImportError
+from composer.utils.object_store.object_store import ObjectStore
+
+__all__ = ['OCIObjectStore']
+
+
+class OCIObjectStore(ObjectStore):
+    """Utility for uploading to and downloading from an OCI bucket.
+
+    Args:
+        bucket (str): The bucket name.
+        prefix (str): A path prefix such as `folder/subfolder/` to prepend to object names. Defaults to ''.
+    """
+
+    def __init__(
+        self,
+        bucket: str,
+        prefix: str = '',
+    ) -> None:
+        try:
+            import oci
+        except ImportError as e:
+            raise MissingConditionalImportError(conda_package='oci',
+                                                extra_deps_group='oci',
+                                                conda_channel='conda-forge') from e
+
+        # Format paths
+        self.bucket = bucket.strip('/')
+        self.prefix = prefix.strip('/')
+        if self.prefix:
+            self.prefix += '/'
+
+        if 'OCI_CONFIG_FILE' in os.environ:
+            config = oci.config.from_file(os.environ['OCI_CONFIG_FILE'])
+        else:
+            config = oci.config.from_file()
+
+        self.client = oci.object_storage.ObjectStorageClient(config=config,
+                                                             retry_strategy=oci.retry.DEFAULT_RETRY_STRATEGY)
+        self.namespace = self.client.get_namespace().data
+        self.upload_manager = oci.object_storage.UploadManager(self.client)
+
+    def get_uri(self, object_name: str) -> str:
+        return f'oci://{self.bucket}/{object_name}'
+
+    def get_object_size(self, object_name: str) -> int:
+        objects = self.client.list_objects(self.namespace, self.bucket, prefix=object_name, fields='size').data.objects
+        relevant_objects = [obj for obj in objects if obj.name == object_name]
+        if relevant_objects:
+            obj = relevant_objects.pop()
+            return obj.size
+        else:
+            raise FileNotFoundError(f'Unable to locate oci://{self.bucket}@{self.namespace}/{object_name}')
+
+    def upload_object(
+        self,
+        object_name: str,
+        filename: Union[str, pathlib.Path],
+        callback: Optional[Callable[[int, int], None]] = None,
+    ):
+        del callback
+
+        self.upload_manager.upload_file(namespace_name=self.namespace,
+                                        bucket_name=self.bucket,
+                                        object_name=object_name,
+                                        file_path=filename)
+
+    def download_object(
+        self,
+        object_name: str,
+        filename: Union[str, pathlib.Path],
+        overwrite: bool = False,
+        callback: Optional[Callable[[int, int], None]] = None,
+    ):
+        del callback
```

Comment:
Is there a reason the logic in this function is not identical to the `S3ObjectStore` logic (modulo the exist API of boto vs oci)?

Reply:
I think `upload_file` for s3 takes a callback and that function oci doesn't

Reply:
ah, i see. there is also some difference in the tmp file cleanup that im not sure if needs to be replicated?

Reply:
right, but we'll keep it simple for now